File: Lectures/Lecture 4/cpp/exercise3.cpp
#include "deque.cpp"

using namespace std;

/**
 * Är strängen "word" ett palindrom
*/
bool is_palindrome(string word) {
    Deque dq = Deque();

    // Fyll vår deque med tecken
    for (int i=0; i<word.length(); i++) {
        dq.add_last(word[i]);
    }

    while (dq.size()>1) {
        // Ta första och sista tecknet från vår deque
        char first = dq.remove_first();
        char last = dq.remove_last();
        if (first != last) {
            return false;
        }
    }

    return true;
}

int main() {
    cout << std::boolalpha;
    cout << "radar:" << is_palindrome("radar") << endl; // True
    cout << "racecar:" << is_palindrome("racecar") << endl; // True
    cout << "tacocat:" << is_palindrome("tacocat") << endl; // True
    cout << "palindrome:" << is_palindrome("palindrome") << endl; // False
}

File: Lectures/Lecture 4/python/exercise3.py
from linked_list import Deque


def is_palindrome(word: str):
    # Är strängen "word" en palindrom

    dq = Deque()

    # Fylla vår deque med tecken
    for letter in word:
        dq.add_last(letter)

    while len(dq) > 1:
        # Ta första och sista tecknet från vår deque
        first = dq.remove_first()
        last = dq.remove_last()
        if first != last:
            return False

    return True


if __name__ == '__main__':
    print(is_palindrome('radar'))  # True
    print(is_palindrome('racecar'))  # True
    print(is_palindrome('tacocat'))  # True
    print(is_palindrome('palindrome'))  # False
